From eae283af2ccd8dc057a6ed52dba6cbd3a15d3629 Mon Sep 17 00:00:00 2001
From: Martin Fick <mfick@codeaurora.org>
Date: Thu, 16 Jun 2011 10:54:36 -0600
Subject: [PATCH] Reuse fileList and more from PatchSreen when switching types

Reuse many of the PatchScreen details when switching from
Side-By-Side to Unified and vice versa.  This can save a lot
of time and CPU when fileLists are large and at least one RPC.

Change-Id: Ibcc58f9c3bac3ebef0b8dc2250d9e0bd1e003b15
---
 .../main/java/com/google/gerrit/client/Gerrit.java |    5 +--
 .../google/gerrit/client/patches/PatchScreen.java  |   12 ++++++++++
 .../com/google/gerrit/client/ui/PatchLink.java     |   23 +++++++++++++++++--
 3 files changed, 34 insertions(+), 6 deletions(-)

diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
index 2dcac32..1d27b80 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/Gerrit.java
@@ -23,6 +23,7 @@ import com.google.gerrit.client.rpc.GerritCallback;
 import com.google.gerrit.client.ui.LinkMenuBar;
 import com.google.gerrit.client.ui.LinkMenuItem;
 import com.google.gerrit.client.ui.MorphingTabPanel;
+import com.google.gerrit.client.ui.PatchLink;
 import com.google.gerrit.client.ui.Screen;
 import com.google.gerrit.common.ClientVersion;
 import com.google.gerrit.common.PageLinks;
@@ -619,9 +620,7 @@ public class Gerrit implements EntryPoint {
               patchScreen.setTopView(PatchScreen.TopView.MAIN);
               AnchorElement.as(getElement()).blur();
             } else {
-              final Patch.Key id = patchScreen.getPatchKey();
-              setTargetHistoryToken(Dispatcher.toPatch(type, id));
-              super.go();
+              new PatchLink("", type, patchScreen).go();
             }
           }
         }
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
index b25b4bc..dcfb34a 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/patches/PatchScreen.java
@@ -331,6 +331,18 @@ public abstract class PatchScreen extends Screen implements
     return patchKey;
   }
 
+  public int getPatchIndex() {
+    return patchIndex;
+  }
+
+  public PatchSetDetail getPatchSetDetail() {
+    return patchSetDetail;
+  }
+
+  public PatchTable getFileList() {
+    return fileList;
+  }
+
   protected void refresh(final boolean isFirst) {
     final int rpcseq = ++rpcSequence;
     lastScript = null;
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java
index 9979edf..7f6de05 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/ui/PatchLink.java
@@ -16,10 +16,11 @@ package com.google.gerrit.client.ui;
 
 import com.google.gerrit.client.Dispatcher;
 import com.google.gerrit.client.changes.PatchTable;
+import com.google.gerrit.client.patches.PatchScreen;
 import com.google.gerrit.common.data.PatchSetDetail;
 import com.google.gerrit.reviewdb.Patch;
 
-public abstract class PatchLink extends InlineHyperlink {
+public class PatchLink extends InlineHyperlink {
   protected Patch.Key patchKey;
   protected int patchIndex;
   protected PatchSetDetail patchSetDetail;
@@ -30,10 +31,10 @@ public abstract class PatchLink extends InlineHyperlink {
    * @param patchKey The key for this patch
    * @param patchIndex The index of the current patch in the patch set
    * @param historyToken The history token
-   * @parma patchSetDetail Detailed information about the patch set.
+   * @param patchSetDetail Detailed information about the patch set.
    * @param parentPatchTable The table used to display this link
    */
-  public PatchLink(final String text, final Patch.Key patchKey,
+  protected PatchLink(final String text, final Patch.Key patchKey,
       final int patchIndex, final String historyToken,
       final PatchSetDetail patchSetDetail,
       final PatchTable parentPatchTable) {
@@ -44,6 +45,22 @@ public abstract class PatchLink extends InlineHyperlink {
     this.parentPatchTable = parentPatchTable;
   }
 
+  /**
+   * @param text The text of this link
+   * @param type The type of the link to create (unified/side-by-side)
+   * @param patchScreen The patchScreen to grab contents to link to from
+   */
+  public PatchLink(final String text, final PatchScreen.Type type,
+      final PatchScreen patchScreen) {
+    this(text, //
+        patchScreen.getPatchKey(), //
+        patchScreen.getPatchIndex(), //
+        Dispatcher.toPatch(type, patchScreen.getPatchKey()), //
+        patchScreen.getPatchSetDetail(), //
+        patchScreen.getFileList() //
+        );
+  }
+
   @Override
   public void go() {
     Dispatcher.patch( //
-- 
1.7.3.1

